### PR TITLE
Scalaz proposal

### LINKED ISSUE #1: scalaz
lots that a contributor can help out with. Some selfish suggestions

- https://github.com/scalaz/scalaz/issues/1662 grunt work with a huge payoff (this lets us do generic derivation of `F[_]` shaped things as per the `MonadErrorTest` example). Suitable for 7.2.
- https://github.com/scalaz/scalaz/issues/1658 for the mathematically inclined, needs to be 7.3+.

### PR DESCRIPTION
**Pull request by zainab-ali**
Closes #1

### CODE REVIEW

**Comment by zainab-ali** [comment]
Let's get a review from @fommil before merging in

**Comment by fommil** [comment]
looks good, although https://github.com/scalaz/scalaz/issues/1618 laws for Codivide and Coapplicative would be more valuable than Reader

**Review by DanielaSfregola — APPROVED** [review]
LGTM

**Pull request merged**
